File: card-core/src/main/java/com/jean202/cardmizer/core/domain/SpendingPeriod.java
package com.jean202.cardmizer.core.domain;

import java.time.LocalDate;
import java.time.YearMonth;
import java.util.List;
import java.util.Objects;
import java.util.stream.IntStream;

public record SpendingPeriod(YearMonth yearMonth) {
    public SpendingPeriod {
        Objects.requireNonNull(yearMonth, "yearMonth must not be null");
    }

    public boolean includes(LocalDate date) {
        Objects.requireNonNull(date, "date must not be null");
        return YearMonth.from(date).equals(yearMonth);
    }

    public SpendingPeriod previous() {
        return new SpendingPeriod(yearMonth.minusMonths(1));
    }

    public List<SpendingPeriod> fromStartOfYear() {
        return IntStream.rangeClosed(1, yearMonth.getMonthValue())
                .mapToObj(month -> new SpendingPeriod(YearMonth.of(yearMonth.getYear(), month)))
                .toList();
    }
}

File: card-core/src/main/kotlin/com/jean202/cardmizer/core/domain/SpendingPeriod.kt
package com.jean202.cardmizer.core.domain

import java.time.LocalDate
import java.time.YearMonth

data class SpendingPeriod(val yearMonth: YearMonth) {

    fun includes(date: LocalDate): Boolean = YearMonth.from(date) == yearMonth

    fun previous(): SpendingPeriod = SpendingPeriod(yearMonth.minusMonths(1))

    fun fromStartOfYear(): List<SpendingPeriod> =
        (1..yearMonth.monthValue).map { month ->
            SpendingPeriod(YearMonth.of(yearMonth.year, month))
        }
}
